File: src/agent1/tools/skill_tool.py
"""skill tool: search/install/read/uninstall (aligned with java SkillTool, requires git for GitHub install)."""

from __future__ import annotations

import json
import re
import shutil
import subprocess
import tempfile
import urllib.error
import urllib.request
from pathlib import Path
from typing import List, Optional

from agent1.skills.loader import ClaudeSkillLoader

GITHUB_SEARCH_API = "https://api.github.com/search/repositories"
DEFAULT_DEST_ROOT = ".claude/skills"
_GITHUB_SLUG = re.compile(r"^[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+$")
_GITHUB_URL = re.compile(r"^https?://github\.com/([A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+)(?:/.*)?$", re.I)
_DEFAULT_SEARCH_LIMIT = 5
_MAX_SEARCH_LIMIT = 10


def make_skill_tool(workspace_root: Path):
    root = workspace_root.resolve()
    loader = ClaudeSkillLoader()

    def skill(
        action: str,
        query: str = "",
        limit: int = _DEFAULT_SEARCH_LIMIT,
        source: str = "",
        skill_name: str = "",
        overwrite: bool = False,
        destination_root: str = DEFAULT_DEST_ROOT,
    ) -> str:
        """Manage Claude-style skills: search|install|read|uninstall.

        Args:
            action: search, install, read, or uninstall.
            query: GitHub search query when action=search.
            limit: Max search results (default 5, max 10).
            source: GitHub owner/repo, URL, or local path under workspace (install).
            skill_name: Target skill for read/uninstall; disambiguate on install.
            overwrite: Overwrite existing skill directory on install.
            destination_root: Install root relative to workspace (default .claude/skills).
        """
        act = (action or "").strip().lower()
        if not act:
            return "错误：action 不能为空，支持 search/install/read/uninstall"
        try:
            if act == "search":
                return _search(query, limit)
            if act == "install":
                return _install(
                    root,
                    source,
                    skill_name,
                    overwrite,
                    destination_root or DEFAULT_DEST_ROOT,
                )
            if act == "read":
                return _read_skill(root, loader, skill_name)
            if act == "uninstall":
                return _uninstall(root, loader, skill_name)
            return f"错误：不支持的 action={action}，支持 search/install/read/uninstall"
        except Exception as e:
            return f"错误：skill 工具执行失败: {e}"

    return skill


def _search(query: str, limit: int) -> str:
    q = (query or "").strip()
    if not q:
        return "错误：action=search 时 query 不能为空"
    lim = _DEFAULT_SEARCH_LIMIT if limit <= 0 else min(int(limit), _MAX_SEARCH_LIMIT)
    github_query = q + " (claude skill OR cursor skill OR SKILL.md) in:name,description,readme"
    from urllib.parse import quote

    url = f"{GITHUB_SEARCH_API}?q={quote(github_query)}&sort=stars&order=desc&per_page={lim}"
    req = urllib.request.Request(
        url,
        headers={
            "Accept": "application/vnd.github+json",
            "X-GitHub-Api-Version": "2022-11-28",
            "User-Agent": "agent1-python-skill-tool",
        },
        method="GET",
    )
    code = 0
    body = ""
    try:
        with urllib.request.urlopen(req, timeout=20) as resp:
            body = resp.read().decode("utf-8", errors="replace")
            code = resp.status
    except urllib.error.HTTPError as e:
        return f"错误：GitHub 搜索失败，status={e.code}"
    except Exception as e:
        return f"错误：GitHub 搜索失败: {e}"
    if code < 200 or code >= 300:
        return f"错误：GitHub 搜索失败，status={code}"
    data = json.loads(body)
    items = data.get("items") or []
    if not items:
        return "没有找到匹配的技能仓库。"
    lines: List[str] = ["搜索结果（可用于 action=install 的 source）:"]
    for idx, item in enumerate(items, start=1):
        full_name = (item.get("full_name") or "").strip()
        if not full_name:
            continue
        html_url = item.get("html_url") or ""
        desc = (item.get("description") or "").replace("\n", " ").strip()
        stars = item.get("stargazers_count") or 0
        lines.append(
            f"{idx}. {full_name} | stars={stars}" + (f" | {desc}" if desc else "") + f"\n   {html_url}"
        )
    lines.append("")
    lines.append('安装示例：skill action=install source="owner/repo"')
    return "\n".join(lines)


def _install(
    workspace: Path,
    source: str,
    requested_skill_name: str,
    overwrite: bool,
    destination_root_raw: str,
) -> str:
    src = (source or "").strip()
    if not src:
        return "错误：action=install 时 source 不能为空"
    dest_root_in = (destination_root_raw or DEFAULT_DEST_ROOT).strip() or DEFAULT_DEST_ROOT
    destination_root = _resolve_within_workspace(workspace, dest_root_in)
    if destination_root is None:
        return "错误：destination_root 超出工作区范围"

    temp_dir: Optional[Path] = None
    try:
        if _looks_like_github(src):
            temp_dir = Path(tempfile.mkdtemp(prefix="agent1_skill_install_"))
            slug = _normalize_github_slug(src)
            branch = _fetch_default_branch(slug)
            source_root = _git_clone(slug, branch, temp_dir / "repo")
        else:
            local = _resolve_within_workspace(workspace, src)
            if local is None:
                return "错误：本地 source 超出工作区范围"
            source_root = local
        if not source_root.is_dir():
            return f"错误：source 不是目录: {source_root}"

        skill_dirs = _detect_skill_directories(source_root)
        if not skill_dirs:
            return "错误：未在 source 中找到 SKILL.md"
        selected = _select_skill_directory(skill_dirs, requested_skill_name.strip())
        if selected is None:
            names = [p.name for p in skill_dirs]
            return "检测到多个 skill，请指定 skill_name: " + ", ".join(names)

        directory_name = _sanitize_directory_name(selected.name)
        if not directory_name:
            return "错误：skill 目录名非法"

        destination_root.mkdir(parents=True, exist_ok=True)
        target_dir = (destination_root / directory_name).resolve()
        if not str(target_dir).startswith(str(destination_root.resolve())):
            return "错误：目标路径非法"
        if target_dir.exists():
            if not overwrite:
                return f"错误：目标 skill 已存在，若要覆盖请设置 overwrite=true: {target_dir}"
            shutil.rmtree(target_dir, ignore_errors=True)
        shutil.copytree(selected, target_dir)
        rel = target_dir.relative_to(workspace)
        return (
            f"安装成功: {directory_name}\n路径: {rel}\n"
            f'可通过 skill(action=read, skill_name="{directory_name}") 读取。'
        )
    finally:
        if temp_dir and temp_dir.exists():
            shutil.rmtree(temp_dir, ignore_errors=True)


def _read_skill(workspace: Path, loader: ClaudeSkillLoader, skill_name: str) -> str:
    name = (skill_name or "").strip()
    if not name:
        return "错误：action=read 时 skill_name 不能为空"
    load_result = loader.load_from_project_root(workspace)
    matched = None
    for sk in load_result.skills:
        dir_name = sk.source_path.parent.name
        if sk.name.lower() == name.lower() or dir_name.lower() == name.lower():
            matched = sk
            break
    if matched is None:
        names = sorted({s.name for s in load_result.skills}, key=str.lower)
        hint = "\n当前未发现任何已安装 skill。" if not names else "\n当前可用 skill: " + ", ".join(names)
        return f"未找到 skill: {name}{hint}"
    parts = [f"skill: {matched.name}"]
    if matched.description:
        parts.append(f"description: {matched.description}")
    if matched.frontmatter:
        parts.append("frontmatter:")
        for k, v in matched.frontmatter.items():
            parts.append(f"- {k}: {v}")
    parts.append("\nSKILL.md content:")
    parts.append(matched.content.strip())
    return "\n".join(parts).strip()


def _uninstall(workspace: Path, loader: ClaudeSkillLoader, skill_name: str) -> str:
    name = (skill_name or "").strip()
    if not name:
        return "错误：action=uninstall 时 skill_name 不能为空"
    skills_root = (workspace / DEFAULT_DEST_ROOT).resolve()
    if not str(skills_root).startswith(str(workspace)):
        return "错误：skills 目录非法"
    if not skills_root.is_dir():
        return "当前没有可卸载的 skill（未找到 .claude/skills 目录）。"
    direct = (skills_root / name).resolve()
    if str(direct).startswith(str(skills_root)) and direct.is_dir():
        shutil.rmtree(direct, ignore_errors=True)
        return f"卸载成功: {name}"
    load_result = loader.load_from_project_root(workspace)
    for sk in load_result.skills:
        dir_name = sk.source_path.parent.name
        if sk.name.lower() == name.lower() or dir_name.lower() == name.lower():
            target = (skills_root / dir_name).resolve()
            if not str(target).startswith(str(skills_root)) or not target.is_dir():
                return f"未找到可卸载目录: {dir_name}"
            shutil.rmtree(target, ignore_errors=True)
            return f"卸载成功: {dir_name}"
    names = sorted({s.name for s in load_result.skills}, key=str.lower)
    hint = "\n当前没有已安装 skill。" if not names else "\n当前可用 skill: " + ", ".join(names)
    return f"未找到 skill: {name}{hint}"


def _resolve_within_workspace(workspace: Path, raw: str) -> Optional[Path]:
    p = Path(raw)
    resolved = p.resolve() if p.is_absolute() else (workspace / p).resolve()
    try:
        resolved.relative_to(workspace)
    except ValueError:
        return None
    return resolved


def _looks_like_github(source: str) -> bool:
    s = source.strip()
    return bool(_GITHUB_SLUG.match(s) or _GITHUB_URL.match(s))


def _normalize_github_slug(source: str) -> str:
    raw = source.strip()
    if _GITHUB_SLUG.match(raw):
        return raw
    m = _GITHUB_URL.match(raw)
    if not m:
        raise ValueError(f"不支持的 GitHub source: {source}")
    return m.group(1)


def _fetch_default_branch(slug: str) -> str:
    url = f"https://api.github.com/repos/{slug}"
    req = urllib.request.Request(
        url,
        headers={
            "Accept": "application/vnd.github+json",
            "X-GitHub-Api-Version": "2022-11-28",
            "User-Agent": "agent1-python-skill-tool",
        },
        method="GET",
    )
    with urllib.request.urlopen(req, timeout=20) as resp:
        data = json.loads(resp.read().decode("utf-8", errors="replace"))
    branch = (data.get("default_branch") or "").strip()
    if not branch:
        raise OSError("无法识别默认分支")
    return branch


def _git_clone(slug: str, branch: str, dest: Path) -> Path:
    dest.parent.mkdir(parents=True, exist_ok=True)
    cmd = [
        "git",
        "clone",
        "--depth",
        "1",
        "--branch",
        branch,
        f"https://github.com/{slug}.git",
        str(dest),
    ]
    proc = subprocess.run(cmd, capture_output=True, text=True, timeout=120)
    if proc.returncode != 0:
        out = (proc.stdout or "") + (proc.stderr or "")
        raise OSError(f"git clone 失败: {out.strip()}")
    return dest


def _detect_skill_directories(source_root: Path) -> List[Path]:
    found: List[Path] = []
    for f in source_root.rglob("SKILL.md"):
        if f.is_file():
            found.append(f.parent)
    return sorted(found, key=lambda p: str(p))


def _select_skill_directory(skill_dirs: List[Path], requested: str) -> Optional[Path]:
    if not skill_dirs:
        return None
    if not requested:
        return skill_dirs[0] if len(skill_dirs) == 1 else None
    for d in skill_dirs:
        if d.name.lower() == requested.lower():
            return d
    return None


def _sanitize_directory_name(name: str) -> str:
    normalized = re.sub(r"[^A-Za-z0-9._-]", "-", name.strip())
    normalized = re.sub(r"-{2,}", "-", normalized)
    return normalized.strip("-")

File: java_agent/cli/src/main/java/com/agent1/javaagent/cli/tools/SkillTool.java
package com.agent1.javaagent.cli.tools;

import com.agent1.javaagent.cli.skills.ClaudeSkill;
import com.agent1.javaagent.cli.skills.ClaudeSkillLoader;
import com.agent1.javaagent.core.CancellationToken;
import com.agent1.javaagent.tool.AgentTool;
import com.agent1.javaagent.tool.ToolExecutionResult;
import com.agent1.javaagent.tool.ToolUpdateListener;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class SkillTool implements AgentTool {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final int DEFAULT_SEARCH_LIMIT = 5;
    private static final int MAX_SEARCH_LIMIT = 10;
    private static final String GITHUB_SEARCH_API = "https://api.github.com/search/repositories";
    private static final String DEFAULT_DEST_ROOT = ".claude/skills";
    private static final Pattern GITHUB_SLUG = Pattern.compile("^[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+$");
    private static final Pattern GITHUB_URL = Pattern.compile("^https?://github\\.com/([A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+)(?:/.*)?$");

    private final Path workspaceRoot;
    private final HttpClient httpClient;
    private final ClaudeSkillLoader skillLoader;

    public SkillTool(Path workspaceRoot) {
        this.workspaceRoot = workspaceRoot.toAbsolutePath().normalize();
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();
        this.skillLoader = new ClaudeSkillLoader();
    }

    @Override
    public String name() {
        return "skill";
    }

    @Override
    public String description() {
        return "Manage skills with action=search|install|read|uninstall.";
    }

    @Override
    public JsonNode parametersSchema() {
        ObjectNode schema = MAPPER.createObjectNode();
        schema.put("type", "object");
        ObjectNode properties = MAPPER.createObjectNode();

        ArrayNode actionEnum = MAPPER.createArrayNode().add("search").add("install").add("read").add("uninstall");
        ObjectNode actionNode = MAPPER.createObjectNode();
        actionNode.put("type", "string");
        actionNode.set("enum", actionEnum);
        actionNode.put("description", "Skill action: search, install, read");
        properties.set("action", actionNode);
        properties.set(
            "query",
            MAPPER.createObjectNode()
                .put("type", "string")
                .put("description", "Search query for action=search")
        );
        properties.set(
            "limit",
            MAPPER.createObjectNode()
                .put("type", "integer")
                .put("description", "Search result count for action=search, default 5, max 10")
        );
        properties.set(
            "source",
            MAPPER.createObjectNode()
                .put("type", "string")
                .put("description", "Source for action=install: GitHub repo (owner/repo or URL) or local dir path")
        );
        properties.set(
            "skill_name",
            MAPPER.createObjectNode()
                .put("type", "string")
                .put("description", "Skill name for action=read/uninstall, or target skill for action=install")
        );
        properties.set(
            "overwrite",
            MAPPER.createObjectNode()
                .put("type", "boolean")
                .put("description", "Overwrite existing installed skill for action=install")
        );
        properties.set(
            "destination_root",
            MAPPER.createObjectNode()
                .put("type", "string")
                .put("description", "Install root under workspace for action=install, default .claude/skills")
        );
        schema.set("properties", properties);
        schema.set("required", MAPPER.createArrayNode().add("action"));
        return schema;
    }

    @Override
    public ToolExecutionResult execute(
        String toolCallId,
        JsonNode parameters,
        CancellationToken cancellationToken,
        ToolUpdateListener onUpdate
    ) {
        String action = parameters.path("action").asText("").trim().toLowerCase(Locale.ROOT);
        if (action.isEmpty()) {
            return ToolExecutionResult.text("错误：action 不能为空，支持 search/install/read/uninstall");
        }
        if (cancellationToken.isCancelled()) {
            return ToolExecutionResult.text("错误：执行已取消");
        }

        try {
            return switch (action) {
                case "search" -> search(parameters, cancellationToken);
                case "install" -> install(parameters, cancellationToken);
                case "read" -> readSkill(parameters, cancellationToken);
                case "uninstall" -> uninstall(parameters, cancellationToken);
                default -> ToolExecutionResult.text("错误：不支持的 action=" + action + "，支持 search/install/read/uninstall");
            };
        } catch (Exception e) {
            return ToolExecutionResult.text("错误：skill 工具执行失败: " + e.getMessage());
        }
    }

    private ToolExecutionResult search(JsonNode parameters, CancellationToken cancellationToken) throws Exception {
        String query = parameters.path("query").asText("").trim();
        if (query.isEmpty()) {
            return ToolExecutionResult.text("错误：action=search 时 query 不能为空");
        }
        if (cancellationToken.isCancelled()) {
            return ToolExecutionResult.text("错误：执行已取消");
        }

        int limit = parameters.path("limit").asInt(DEFAULT_SEARCH_LIMIT);
        if (limit <= 0) {
            limit = DEFAULT_SEARCH_LIMIT;
        }
        limit = Math.min(limit, MAX_SEARCH_LIMIT);

        String githubQuery = query + " (claude skill OR cursor skill OR SKILL.md) in:name,description,readme";
        String encodedQuery = URLEncoder.encode(githubQuery, StandardCharsets.UTF_8);
        URI uri = URI.create(GITHUB_SEARCH_API + "?q=" + encodedQuery + "&sort=stars&order=desc&per_page=" + limit);

        HttpRequest request = HttpRequest.newBuilder(uri)
            .timeout(Duration.ofSeconds(20))
            .header("Accept", "application/vnd.github+json")
            .header("X-GitHub-Api-Version", "2022-11-28")
            .header("User-Agent", "java-agent-skill-tool")
            .GET()
            .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            return ToolExecutionResult.text("错误：GitHub 搜索失败，status=" + response.statusCode());
        }

        JsonNode root = MAPPER.readTree(response.body());
        JsonNode items = root.path("items");
        if (!items.isArray() || items.isEmpty()) {
            return ToolExecutionResult.text("没有找到匹配的技能仓库。");
        }

        List<String> lines = new ArrayList<>();
        lines.add("搜索结果（可用于 action=install 的 source）:");
        int idx = 1;
        for (JsonNode item : items) {
            String fullName = item.path("full_name").asText("");
            String htmlUrl = item.path("html_url").asText("");
            String description = item.path("description").asText("");
            int stars = item.path("stargazers_count").asInt(0);
            if (fullName.isBlank()) {
                continue;
            }
            lines.add(
                idx + ". " + fullName
                    + " | stars=" + stars
                    + (description.isBlank() ? "" : " | " + description.replace('\n', ' ').trim())
                    + "\n   " + htmlUrl
            );
            idx++;
        }
        if (idx == 1) {
            return ToolExecutionResult.text("没有找到匹配的技能仓库。");
        }
        lines.add("");
        lines.add("安装示例：skill {\"action\":\"install\",\"source\":\"owner/repo\"}");
        return ToolExecutionResult.text(String.join("\n", lines));
    }

    private ToolExecutionResult install(JsonNode parameters, CancellationToken cancellationToken) throws Exception {
        String source = parameters.path("source").asText("").trim();
        if (source.isEmpty()) {
            return ToolExecutionResult.text("错误：action=install 时 source 不能为空");
        }
        if (cancellationToken.isCancelled()) {
            return ToolExecutionResult.text("错误：执行已取消");
        }

        String requestedSkillName = parameters.path("skill_name").asText("").trim();
        boolean overwrite = parameters.path("overwrite").asBoolean(false);
        String destinationRootRaw = parameters.path("destination_root").asText(DEFAULT_DEST_ROOT).trim();
        if (destinationRootRaw.isEmpty()) {
            destinationRootRaw = DEFAULT_DEST_ROOT;
        }
        Path destinationRoot = resolveWithinWorkspace(destinationRootRaw);
        if (destinationRoot == null) {
            return ToolExecutionResult.text("错误：destination_root 超出工作区范围");
        }

        Path tempDir = null;
        try {
            Path sourceRoot;
            if (looksLikeGitHub(source)) {
                tempDir = Files.createTempDirectory("java_agent_skill_install_");
                String repoSlug = normalizeGitHubSlug(source);
                String defaultBranch = fetchDefaultBranch(repoSlug);
                sourceRoot = cloneRepo(repoSlug, defaultBranch, tempDir);
            } else {
                Path local = resolveWithinWorkspace(source);
                if (local == null) {
                    return ToolExecutionResult.text("错误：本地 source 超出工作区范围");
                }
                sourceRoot = local;
            }

            if (!Files.isDirectory(sourceRoot)) {
                return ToolExecutionResult.text("错误：source 不是目录: " + sourceRoot);
            }

            List<Path> skillDirs = detectSkillDirectories(sourceRoot);
            if (skillDirs.isEmpty()) {
                return ToolExecutionResult.text("错误：未在 source 中找到 SKILL.md");
            }

            Path selected = selectSkillDirectory(skillDirs, requestedSkillName);
            if (selected == null) {
                List<String> names = new ArrayList<>();
                for (Path p : skillDirs) {
                    names.add(p.getFileName().toString());
                }
                return ToolExecutionResult.text("检测到多个 skill，请指定 skill_name: " + String.join(", ", names));
            }

            String directoryName = sanitizeDirectoryName(selected.getFileName().toString());
            if (directoryName.isBlank()) {
                return ToolExecutionResult.text("错误：skill 目录名非法");
            }

            Files.createDirectories(destinationRoot);
            Path targetDir = destinationRoot.resolve(directoryName).normalize();
            if (!targetDir.startsWith(destinationRoot)) {
                return ToolExecutionResult.text("错误：目标路径非法");
            }
            if (Files.exists(targetDir)) {
                if (!overwrite) {
                    return ToolExecutionResult.text("错误：目标 skill 已存在，若要覆盖请设置 overwrite=true: " + targetDir);
                }
                deleteRecursively(targetDir);
            }
            copyRecursively(selected, targetDir);

            String relTarget = workspaceRoot.relativize(targetDir).toString();
            return ToolExecutionResult.text(
                "安装成功: " + directoryName
                    + "\n路径: " + relTarget
                    + "\n可通过 skill(action=read, skill_name=\"" + directoryName + "\") 读取。"
            );
        } finally {
            if (tempDir != null) {
                try {
                    deleteRecursively(tempDir);
                } catch (IOException ignored) {
                    // best effort cleanup
                }
            }
        }
    }

    private ToolExecutionResult readSkill(JsonNode parameters, CancellationToken cancellationToken) {
        String skillName = parameters.path("skill_name").asText("").trim();
        if (skillName.isEmpty()) {
            return ToolExecutionResult.text("错误：action=read 时 skill_name 不能为空");
        }
        if (cancellationToken.isCancelled()) {
            return ToolExecutionResult.text("错误：执行已取消");
        }

        ClaudeSkillLoader.SkillLoadResult loadResult = skillLoader.loadFromProjectRoot(workspaceRoot);
        if (!loadResult.warnings().isEmpty()) {
            // warnings are non-fatal; continue to search best effort
        }

        ClaudeSkill matched = null;
        for (ClaudeSkill skill : loadResult.skills()) {
            String dirName = skill.getSourcePath().getParent().getFileName().toString();
            if (skill.getName().equalsIgnoreCase(skillName) || dirName.equalsIgnoreCase(skillName)) {
                matched = skill;
                break;
            }
        }

        if (matched == null) {
            List<String> names = new ArrayList<>();
            for (ClaudeSkill skill : loadResult.skills()) {
                names.add(skill.getName());
            }
            names.sort(String::compareToIgnoreCase);
            return ToolExecutionResult.text(
                "未找到 skill: " + skillName
                    + (names.isEmpty() ? "\n当前未发现任何已安装 skill。" : "\n当前可用 skill: " + String.join(", ", names))
            );
        }

        StringBuilder out = new StringBuilder();
        out.append("skill: ").append(matched.getName()).append("\n");
        if (!matched.getDescription().isBlank()) {
            out.append("description: ").append(matched.getDescription()).append("\n");
        }
        if (!matched.getFrontmatter().isEmpty()) {
            out.append("frontmatter:\n");
            for (Map.Entry<String, String> entry : matched.getFrontmatter().entrySet()) {
                out.append("- ").append(entry.getKey()).append(": ").append(entry.getValue()).append("\n");
            }
        }
        out.append("\nSKILL.md content:\n");
        out.append(matched.getContent().strip());
        return ToolExecutionResult.text(out.toString().trim());
    }

    private ToolExecutionResult uninstall(JsonNode parameters, CancellationToken cancellationToken) throws IOException {
        String skillName = parameters.path("skill_name").asText("").trim();
        if (skillName.isEmpty()) {
            return ToolExecutionResult.text("错误：action=uninstall 时 skill_name 不能为空");
        }
        if (cancellationToken.isCancelled()) {
            return ToolExecutionResult.text("错误：执行已取消");
        }

        Path skillsRoot = workspaceRoot.resolve(DEFAULT_DEST_ROOT).normalize();
        if (!skillsRoot.startsWith(workspaceRoot)) {
            return ToolExecutionResult.text("错误：skills 目录非法");
        }
        if (!Files.isDirectory(skillsRoot)) {
            return ToolExecutionResult.text("当前没有可卸载的 skill（未找到 .claude/skills 目录）。");
        }

        Path direct = skillsRoot.resolve(skillName).normalize();
        if (direct.startsWith(skillsRoot) && Files.isDirectory(direct)) {
            deleteRecursively(direct);
            return ToolExecutionResult.text("卸载成功: " + skillName);
        }

        ClaudeSkillLoader.SkillLoadResult loadResult = skillLoader.loadFromProjectRoot(workspaceRoot);
        for (ClaudeSkill skill : loadResult.skills()) {
            String name = skill.getName();
            String dirName = skill.getSourcePath().getParent().getFileName().toString();
            if (name.equalsIgnoreCase(skillName) || dirName.equalsIgnoreCase(skillName)) {
                Path target = skillsRoot.resolve(dirName).normalize();
                if (!target.startsWith(skillsRoot) || !Files.isDirectory(target)) {
                    return ToolExecutionResult.text("未找到可卸载目录: " + dirName);
                }
                deleteRecursively(target);
                return ToolExecutionResult.text("卸载成功: " + dirName);
            }
        }

        List<String> names = new ArrayList<>();
        for (ClaudeSkill skill : loadResult.skills()) {
            names.add(skill.getName());
        }
        names.sort(String::compareToIgnoreCase);
        return ToolExecutionResult.text(
            "未找到 skill: " + skillName
                + (names.isEmpty() ? "\n当前没有已安装 skill。" : "\n当前可用 skill: " + String.join(", ", names))
        );
    }

    private Path resolveWithinWorkspace(String raw) {
        Path input = Path.of(raw);
        Path resolved = input.isAbsolute()
            ? input.normalize()
            : workspaceRoot.resolve(input).normalize();
        if (!resolved.startsWith(workspaceRoot)) {
            return null;
        }
        return resolved;
    }

    private boolean looksLikeGitHub(String source) {
        return GITHUB_SLUG.matcher(source).matches() || GITHUB_URL.matcher(source).matches();
    }

    private String normalizeGitHubSlug(String source) {
        String raw = source.trim();
        if (GITHUB_SLUG.matcher(raw).matches()) {
            return raw;
        }
        Matcher matcher = GITHUB_URL.matcher(raw);
        if (!matcher.matches()) {
            throw new IllegalArgumentException("不支持的 GitHub source: " + source);
        }
        return matcher.group(1);
    }

    private String fetchDefaultBranch(String slug) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(URI.create("https://api.github.com/repos/" + slug))
            .timeout(Duration.ofSeconds(20))
            .header("Accept", "application/vnd.github+json")
            .header("X-GitHub-Api-Version", "2022-11-28")
            .header("User-Agent", "java-agent-skill-tool")
            .GET()
            .build();
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new IOException("GitHub 仓库信息获取失败，status=" + response.statusCode());
        }
        JsonNode root = MAPPER.readTree(response.body());
        String branch = root.path("default_branch").asText("").trim();
        if (branch.isEmpty()) {
            throw new IOException("无法识别默认分支");
        }
        return branch;
    }

    private Path cloneRepo(String slug, String branch, Path tempDir) throws IOException, InterruptedException {
        Path checkoutDir = tempDir.resolve("repo");
        List<String> command = List.of(
            "git",
            "clone",
            "--depth",
            "1",
            "--branch",
            branch,
            "https://github.com/" + slug + ".git",
            checkoutDir.toString()
        );
        ProcessBuilder pb = new ProcessBuilder(command);
        pb.redirectErrorStream(true);
        Process process = pb.start();
        String output = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
        int exit = process.waitFor();
        if (exit != 0) {
            throw new IOException("git clone 失败: " + output.trim());
        }
        return checkoutDir;
    }

    private List<Path> detectSkillDirectories(Path sourceRoot) throws IOException {
        List<Path> skillDirs = new ArrayList<>();
        try (var walk = Files.walk(sourceRoot)) {
            walk.filter(path -> Files.isRegularFile(path) && "SKILL.md".equals(path.getFileName().toString()))
                .forEach(skillFile -> skillDirs.add(skillFile.getParent()));
        }
        skillDirs.sort(Comparator.comparing(Path::toString));
        return skillDirs;
    }

    private Path selectSkillDirectory(List<Path> skillDirs, String requestedSkillName) {
        if (skillDirs.isEmpty()) {
            return null;
        }
        if (requestedSkillName == null || requestedSkillName.isBlank()) {
            return skillDirs.size() == 1 ? skillDirs.get(0) : null;
        }
        for (Path skillDir : skillDirs) {
            String dirName = skillDir.getFileName().toString();
            if (dirName.equalsIgnoreCase(requestedSkillName.trim())) {
                return skillDir;
            }
        }
        return null;
    }

    private String sanitizeDirectoryName(String name) {
        String normalized = name.trim().replaceAll("[^A-Za-z0-9._-]", "-");
        normalized = normalized.replaceAll("-{2,}", "-");
        return normalized.replaceAll("^-+|-+$", "");
    }

    private void copyRecursively(Path source, Path target) throws IOException {
        try (var walk = Files.walk(source)) {
            for (Path src : (Iterable<Path>) walk::iterator) {
                Path rel = source.relativize(src);
                Path dst = target.resolve(rel.toString());
                if (Files.isDirectory(src)) {
                    Files.createDirectories(dst);
                } else if (Files.isRegularFile(src)) {
                    Files.createDirectories(dst.getParent());
                    Files.copy(src, dst, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.COPY_ATTRIBUTES);
                }
            }
        }
    }

    private void deleteRecursively(Path root) throws IOException {
        if (root == null || !Files.exists(root)) {
            return;
        }
        try (var walk = Files.walk(root)) {
            walk.sorted(Comparator.reverseOrder()).forEach(path -> {
                try {
                    Files.deleteIfExists(path);
                } catch (IOException ignored) {
                    // best effort cleanup
                }
            });
        }
    }
}
